### PR TITLE
Add hard mode

### PR DESCRIPTION
**Pull request by Kudbettin**
Adds a hard mode to the game, with the respective flag.

Restarts the level whenever an incorrect key is pressed.

usage:
`wpm --hard`

### CODE REVIEW

File: wpm/game.py [hunk 1]
```diff
@@ -234,14 +234,14 @@ def handle_key(self, key):
         # Browse mode
         if self.start is None or (self.start is not None and
                                   self.stop is not None):
-            if key in (" ", "KEY_LEFT", "KEY_RIGHT"):
+            if key in ("\t", "KEY_LEFT", "KEY_RIGHT"):
```
Comment (cslarsen) [inline wpm/game.py]: Please put the tab in a separate PR if you think it should be there.

File: wpm/game.py [hunk 2]
```diff
@@ -234,14 +234,14 @@ def handle_key(self, key):
...
                 self.reset(direction=-1 if key == "KEY_LEFT" else 1)
                 return
             elif Screen.is_escape(key):
                 # Exit program
                 raise KeyboardInterrupt()
 
-        if Screen.is_escape(key):
+        if Screen.is_escape(key) or (hard_flag and self.position < len(self.quote.text) and self.quote.text[self.position] != key):
```
Comment (cslarsen) [inline wpm/game.py]: I have a feeling there must be a much better way. If you look where I update `self.incorrect`, I think you have the logic there already; then you just need to react on your flag.

File: wpm/game.py [hunk 3]
```diff
@@ -221,7 +221,7 @@ def resize(self):
                 for inc in range(self.incorrect + 1):
                     self.screen.highlight_progress(self.position, inc)
 
-    def handle_key(self, key):
+    def handle_key(self, key, hard_flag):
```
Comment (cslarsen) [inline wpm/game.py]: Probably better to set the option on some other object (`self`?)

File: wpm/commandline.py [hunk 4]
```diff
@@ -275,7 +278,7 @@ def main():
     try:
         with wpm.game.GameManager(quotes, stats, opts.cpm) as gm:
             try:
-                gm.run(to_front=text_ids)
+                gm.run(to_front=text_ids, hard_flag=opts.hard)
```
Comment (cslarsen) [inline wpm/commandline.py]: As stated below, I think it's better to just store this in the GameManager instance than to pass it around.

**Review by cslarsen — CHANGES_REQUESTED** [review]
Sorry for the long wait. Thanks for the PR. If you fix the above, I will reconsider pulling this PR.

**Comment by cslarsen** [comment]
Also, the name `Add files via upload`? :)

**Comment by Kudbettin** [comment]
Hello!

I agree that changes were quite selfish and ugly. I believe I have fixed them.

This PR might had been my first time hence the name :)

File: wpm/game.py (hunk 1 above)
Comment (Kudbettin) [reply wpm/game.py]: Removed. The reason for having it was that I was often accidentally pressing spacebar right after the game reset due to a typo.

I agree that this should not be changed.

File: wpm/game.py [hunk 5]
```diff
@@ -290,6 +291,8 @@ def handle_key(self, key):
             # Finished typing?
             if self.position == len(self.quote.text):
                 self.mark_finished()
+        elif self.hard_flag:
+            self.reset()
```
Comment (cslarsen) [inline wpm/game.py:332]: Does this work properly with backspace and ESC? I just see that the incorrect counter is updated below in the other IF block. Perhaps that's where it should be? (I haven't _tried_ your PR, so you may be right, I'm just asking if everything else works)

**Review by cslarsen — CHANGES_REQUESTED** [review]
This is much better!

You need to resolve conflicts with recent changes. I'll definitely merge this PR if you fix the conflicts and also look into the comment I have on the placement (testing, basically).

File: wpm/game.py (hunk 5 above)
Comment (Kudbettin) [reply wpm/game.py:332]: Just tested both with hard flag and without the hard flag once more to see ESC and backspace behavior.

both versions have:
- ESC resets the level if playing, exits the program if not playing.
- Backspace allows deleting the current word but won't go past (or before) the current word.
- Meta-backspace resets the level.

during the hard mode the second elif statement is never reached.
- the if block accepts the correct keystrokes
- elif hard.flag accepts everything else

during the regular mode
- the if block accepts the correct keystrokes
- elif hard.flag is always false so the next elif block is run.

I would like to think everything else works with it.

I will resolve the conflicts soon.

**Comment by Kudbettin** [comment]
A friendly bump :)

**Comment by Kudbettin** [comment]
I forgot to branch out before making another feature (redlist).

I will close this pr and open one for each.
